### ISSUE TITLE
Dictionary<TKey, TValue>.KeyCollection and Dictionary<TKey, TValue>.ValueCollection are missing

### ISSUE THREAD

**Issue by TriEdgeAI** [bridgedotnet/Bridge#3004]

**Comment by hardhub** [comment]
Why do you need those collections?

Keys and Values collections are available.

**Comment by TriEdgeAI** [comment]
We are trying to port an existing codebase to HTML5 which uses them.

**Comment by hardhub** [comment]
I guess first one can be replaced by HashSet<T>, second one by List<T>.
I do not know exact purpose but maybe you need to create your own KeyCollection / ValueCollection classes derived from HashSet<T> / List<T>..

**Comment by TriEdgeAI** [comment]
Any ETA on this?

**Comment by fabriciomurta** [comment]
This is most likely coming with #3223. As the interfaces are developed, the implementations will require the interface methods.

Actually, maybe this should be assumed as a duplicate to #3223 (which, by the way, is quite ample).

**Comment by TriEdgeAI** [comment]
Nice!

**Comment by TriEdgeAI** [comment]
Isn't this pretty much done now?

**Comment by cshtml5** [comment]
We just posted a [$100 bounty](https://www.bountysource.com/issues/47713606-dictionary-tkey-tvalue-keycollection-and-dictionary-tkey-tvalue-valuecollection-are-missing) to encourage the community to contribute to this feature. Other backers can also add bounties by following the link in the previous sentence.

Please note that, in addition to the "Values" and "Keys" properties of the Dictionary, we also need the ".**GetEnumerator**" method of the Dictionary to have the exact same signature as the one in the .NET Framework (cf. [the lack of **IDictionaryEnumerator** mentioned in issue 3223](https://github.com/bridgedotnet/Bridge/issues/3223#issuecomment-363956461)).

Thank you so much!

**Comment by geoffreymcgill** [comment]
@TriEdgeAI If you have a fix, please send a PR.

**Comment by cshtml5** [comment]
@TriEdgeAI Yes, sure! It doesn't matter if you reported the issue in the first place :)

**Comment by TriEdgeAI** [comment]
Okay, I'll make a PR by the end of the week.

**Comment by TriEdgeAI** [comment]
Yeah, we had to implement it 1:1 as is in .NET Framework because another company wanted to use our software with their existing codebase which referenced third-party libraries which relied on Dictionary<> and all of it's nested classes. So the only way to get it all to work was to have perfect support for Dictionary<>. You are just catching me in a really busy moment and I need to shed 1-2 hours to go through our code and take out what we did, verify it and make a pull-request. I hope to be able to do it by the end of the week.

**Comment by cshtml5** [comment]
@TriEdgeAI Awesome! Thank you very much.

**Comment by cshtml5** [comment]
@TriEdgeAI Hi! By any chance do you think you will have time before Tuesday January 1st to do the pull-request? It's because it is a missing piece to make our integration work. Thanks a lot! Season's greetings.

**Comment by TriEdgeAI** [comment]
Yeah, I'll do it tomorrow. Promise :D

**Comment by cshtml5** [comment]
@TriEdgeAI Happy New Year! Any ETA for the PR? Because it's really critical to what we are doing. Thanks!

**Comment by TriEdgeAI** [comment]
Working on it right now.

**Comment by TriEdgeAI** [comment]
A'ight. Try this.

**Comment by cshtml5** [comment]
@TriEdgeAI Thank you very much! We are going to test it. I will keep you updated over the next few days.

**Comment by cshtml5** [comment]
Hi!

We have actually found an error when attempting to call the GetEnumerator() method using the Pull Request 3818: https://github.com/bridgedotnet/Bridge/pull/3818

### Here is the code to reproduce:

```
Dictionary<string, string> dic1 = new Dictionary<string, string>();
dic1.Add("item1", "the item 1");
dic1.Add("item2", "the item 2");
dic1.Add("item3", "the item 3");
var enumerator = dic1.GetEnumerator();
```

### Expected result:
No error

### Actual result:
The following JavaScript error appears:

`TypeError: dic1.getEnumerator is not a function`

![dictionary-getenumerator-error](https://user-images.githubusercontent.com/17248999/52122376-1506f880-2623-11e9-92c4-acaf73d3c42d.png)



### Note: the error disappears if I cast to IEnumerable<KeyValuePair<string, string>

Just for information, when casting to IEnumerable<KeyValuePair<string, string> the error disappears:
`var enumerator = ((IEnumerable<KeyValuePair<string, string>>)dic1).GetEnumerator();`

![dictionary-getenumerator-workaround](https://user-images.githubusercontent.com/17248999/52122582-c017b200-2623-11e9-96ab-bee859be8073.png)


Any idea how this issue could be solved?

Thank you!

**Comment by fabriciomurta** [comment]
Fixed and merged via #3841. We didn't have @TriEdgeAI sign the CLA so we weren't (policy-wise) allowed to accept his pull request in time, unfortunately. But @vladsch implementation comes along with lots of unit tests and as far as I could see, even more features and meshup of the classes, so this should be a good one.

**Comment by fabriciomurta** [comment]
Ops, the fix will be available next Bridge release -- see attached milestone for version it should be included in!
